frontend/cypress/integration/common/app_details.ts

```diff
@@ -0,0 +1,71 @@
+import { Then } from 'cypress-cucumber-preprocessor/steps';
+
+const APP = 'details';
+const NAMESPACE = 'bookinfo';
+
+Then('user sees details information for app', () => {
+  cy.getBySel('app-description-card').within(() => {
+    cy.get('#pfbadge-A').parent().parent().contains('details'); // App
+    cy.get('#pfbadge-W').parent().parent().contains('details-v1'); // Workload
+    cy.get('#pfbadge-S').parent().parent().contains('details'); // Service
+  });
+});
+
+Then('user sees a minigraph', () => {
+  cy.getBySel('mini-graph');
+});
+
+function openTab(tab: string) {
+  cy.get('#basic-tabs').should('be.visible').contains(tab).click();
+}
+
+Then('user sees inbound and outbound traffic information', () => {
+  openTab('Traffic');
+  cy.contains('Inbound Traffic');
+  // Details app should have some inbound traffic.
+  cy.contains('No Inbound Traffic').should('not.exist');
+  cy.contains('Outbound Traffic');
+});
+
+Then('user sees inbound metrics information', () => {
+  cy.intercept(Cypress.config('baseUrl') + `/api/namespaces/${NAMESPACE}/apps/${APP}/dashboard*`).as('fetchMetrics');
+  openTab('Inbound Metrics');
+  cy.wait('@fetchMetrics');
+  cy.waitForReact(1000, '#root');
+  cy.getReact('IstioMetrics', { props: { 'data-test': 'inbound-metrics-component' } })
+    // Theres be multiple components of the same name providing state through redux/hooks.
+    // This gets us the one we we want but this tests will fail if the context providers
+    // change names or are removed.
+    .nthNode(2)
+    .getCurrentState()
+    .then(state => {
+      cy.wrap(state.dashboard).should('not.be.empty');
+    });
+});
+
+Then('user sees outbound metrics information', () => {
+  cy.intercept(Cypress.config('baseUrl') + `/api/namespaces/${NAMESPACE}/apps/${APP}/dashboard*`).as('fetchMetrics');
+  openTab('Outbound Metrics');
+  cy.wait('@fetchMetrics');
+  cy.waitForReact(1000, '#root');
+  cy.getReact('IstioMetrics', { props: { 'data-test': 'outbound-metrics-component' } })
```

Comment:
Just to confirm, you are grabbing a component, getting its state and then asserting the values of for example, the dashboard.

Maybe is for a followup PR but do we need to test this component checking of the information is correctly displayed right?.

Reply:
Yes I can add some assertions about a chart being rendered but since the actual elements in that chart are just svgs, I couldn't think of a good way to tell if these had displayed correctly. So I just test that the component has the correct state which at least tests that the jaeger integration is working properly. If you can think of a way to test that the charts have rendered correctly I'm happy to update. The tests are pretty sparse though. I can at least add tests for interacting with the "display" menu.

Reply:
Yes, I understand, and it's true. Well, this comment only applies for other tests where it's more suitable, but of course you won't be testing the state in those.